test/e2e/e2ethanos/services.go

```diff
@@ -0,0 +1,315 @@
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/cortexproject/cortex/integration/e2e"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/relabel"
+	"github.com/thanos-io/thanos/pkg/alert"
+	"github.com/thanos-io/thanos/pkg/objstore/client"
+	"github.com/thanos-io/thanos/pkg/query"
+	"github.com/thanos-io/thanos/pkg/receive"
+	"gopkg.in/yaml.v2"
+)
+
+const logLevel = "info"
+
+// TODO(bwplotka): Run against multiple?
+func DefaultPrometheusImage() string {
+	return "quay.io/prometheus/prometheus:v2.16.0"
+}
+
+func DefaultAlertmanagerImage() string {
+	return "quay.io/prometheus/alertmanager:v0.20.0"
+}
+
+// DefaultImage returns the local docker image to use to run Thanos.
+func DefaultImage() string {
+	// Get the Thanos image from the THANOS_IMAGE env variable.
+	if os.Getenv("THANOS_IMAGE") != "" {
+		return os.Getenv("THANOS_IMAGE")
+	}
+
+	return "thanos"
+}
+
+func NewPrometheus(sharedDir string, name string, config, promImage string) (*e2e.HTTPService, string, error) {
+	dir := filepath.Join(sharedDir, "data", "prometheus", name)
+	container := filepath.Join(e2e.ContainerSharedDir, "data", "prometheus", name)
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return nil, "", errors.Wrap(err, "create prometheus dir")
+	}
+
+	if err := ioutil.WriteFile(filepath.Join(dir, "prometheus.yml"), []byte(config), 0666); err != nil {
+		return nil, "", errors.Wrap(err, "creating prom config failed")
+	}
+
+	prom := e2e.NewHTTPService(
+		fmt.Sprintf("prometheus-%s", name),
+		promImage,
+		e2e.NewCommandWithoutEntrypoint("prometheus", e2e.BuildArgs(map[string]string{
+			"--config.file":                     filepath.Join(container, "prometheus.yml"),
+			"--storage.tsdb.path":               filepath.Join(container),
+			"--storage.tsdb.max-block-duration": "2h",
+			"--log.level":                       logLevel,
+			"--web.listen-address":              ":9090",
+		})...),
+		e2e.NewReadinessProbe(9090, "/-/ready", 200),
+		9090,
+	)
+	prom.SetUser("root")
+
+	return prom, container, nil
+}
+
+func NewPrometheusWithSidecar(sharedDir string, netName string, name string, config, promImage string) (*e2e.HTTPService, *Service, error) {
+	prom, dataDir, err := NewPrometheus(sharedDir, name, config, promImage)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sidecar := NewService(
+		fmt.Sprintf("sidecar-%s", name),
+		DefaultImage(),
+		e2e.NewCommand("sidecar", e2e.BuildArgs(map[string]string{
+			"--debug.name":        fmt.Sprintf("sidecar-%v", name),
+			"--grpc-address":      ":9091",
+			"--grpc-grace-period": "0s",
+			"--http-address":      ":80",
+			"--prometheus.url":    "http://" + prom.NetworkEndpointFor(netName, 9090),
+			"--tsdb.path":         filepath.Join(dataDir),
+			"--log.level":         logLevel,
+		})...),
+		e2e.NewReadinessProbe(80, "/-/ready", 200),
+		80,
+		9091,
+	)
+	return prom, sidecar, nil
+}
+
+func generateFileSD(addresses []string) string {
```

Comment:
I guess this is easier than constructing the whole thing with a proper type :smile:

Reply:
:+1: